packages/contracts-sdk/src/implementations/farcaster/FarcasterKeyGatewayContract.ts

```diff
@@ -0,0 +1,248 @@
+        keyType: 1,
+        key: keyToAdd,
+        metadataType: 1,
+        metadata: encodedMetadata,
+        nonce: latestNonce,
+        deadline: deadline,
+      };
+
+      const signer = this.providerOrSigner as ethers.Signer;
+
+      return ResultAsync.fromPromise(
+        signer.signTypedData(
+          KEY_GATEWAY_EIP_712_TYPES.domain,
+          this.removeReadonlyFromReadonlyTypes(KEY_GATEWAY_EIP_712_TYPES.types),
+          addKeyMessage,
+        ) as Promise<SignedAddSignature>,
+        (e) => {
+          return e as FarcasterKeyGatewayContractError;
+        },
+      );
+    });
+  }
+
+  public getSignedKeyRequestSignatureAndEncodedMetadata(
+    ownerFid: FarcasterId,
+    keyToAdd: EVMAccountAddress, // key to be tied to account
+    deadline: UnixTimestamp,
+  ): ResultAsync<SignedKeyRequest, FarcasterKeyGatewayContractError> {
+    return this.getSignedKeyRequestSignature(
+      ownerFid,
+      keyToAdd,
+      deadline,
+    ).andThen((metadataSignature) => {
+      const metadataStruct = {
+        requestFid: ownerFid,
+        requestSigner: keyToAdd,
+        signature: metadataSignature,
+        deadline: deadline,
+      };
+
+      const metadataStructType = ParamType.from({
+        components: [
+          {
+            name: "requestFid",
+            type: "uint256",
+          },
+          {
+            name: "requestSigner",
+            type: "address",
+          },
+          {
+            name: "signature",
+            type: "bytes",
+          },
+          {
+            name: "deadline",
+            type: "uint256",
+          },
+        ],
+        name: "SignedKeyRequestMetadata",
+        type: "tuple",
+      });
+
+      const encodedMetadataStruct = ethers.AbiCoder.defaultAbiCoder().encode(
+        [metadataStructType],
+        [metadataStruct],
+      );
+
+      return okAsync(
+        new SignedKeyRequest(
+          SignedKeyRequestSignature(metadataSignature),
+          EncodedSignedKeyRequestMetadata(encodedMetadataStruct),
+        ),
+      );
+    });
+  }
+
+  // Returns SignedKeyRequestSignature
+  public getSignedKeyRequestSignature(
+    ownerFid: FarcasterId,
+    keyToAdd: EVMAccountAddress, // key to be tied to account
+    deadline: UnixTimestamp,
+  ): ResultAsync<SignedKeyRequestSignature, FarcasterKeyGatewayContractError> {
+    //    https://docs.farcaster.xyz/reference/contracts/reference/signed-key-request-validator#signedkeyrequestmetadata-struct
+    //    Needs to be a EIP712 signature
+    const metadataMessage = {
+      requestFid: ownerFid,
+      key: keyToAdd,
+      deadline,
+    };
+
+    // Use the wallet of the sdk caller as the signer
+    const signer = this.providerOrSigner as ethers.Signer;
+
+    return ResultAsync.fromPromise(
+      signer.signTypedData(
+        SIGNED_KEY_REQUEST_VALIDATOR_EIP_712_TYPES.domain,
+        this.removeReadonlyFromReadonlyTypes(
+          SIGNED_KEY_REQUEST_VALIDATOR_EIP_712_TYPES.types,
```

Comment:
Just a heads up, I think that these types, provided by the hubs package, are setup for Viem, and they may not be correctly formatted for Ethers. You should check the format of the types objects we use elsewhere in the protocol, and compare to what we're getting. removeReadonlyFromReadonlyTypes() is only doing a type cast; that's fine if the underlying type is the same as what Ethers wants.